lib/Dialect/Torch/IR/TorchTypes.cpp

```diff
@@ -372,6 +372,10 @@ static Type convertDtypeToBuiltinElementType(MLIRContext *context, Type dtype) {
   } else if (auto integerType = dtype.dyn_cast<IntegerType>()) {
     return IntegerType::get(context, integerType.getWidth(),
                             IntegerType::Signless);
+  } else if (auto integerType = dtype.dyn_cast<QInt8Type>()) {
+    return IntegerType::get(context, 8, IntegerType::Signless);
```

Comment:
This is not correct. As discussed in #1677, a tensor with !torch.qint8 dtype needs to expand to 3 separate values -- the tensor, scale, and zero point. We want to get this right from the very beginning to allow incremental progress. I think that it would make sense to include a quantized add in this PR as well since that exercises this.

Reply:
I was torn on this question. Just having quantize_per_tensor + int_repr is useful to the networks that we are looking at right now. I didn't want to delay this PR by making it bigger with the quantized add.
My initial plan was to only add scale and zero point together with quantized add in another PR.

Would you agree to changing the type to (tensor, scale, zero point) in this PR, but delay quantized add to another PR?

Reply:
I think we need to implement quantized add end-to-end in the first PR. There are other things besides just converting the types that we want to get right from the start (materializations, file structure for quantized operators, etc.). I don't think we will have full visibility of the consequences of changing the type to (tensor, scale, zero point) without implementing quantized add end-to-end.

Reply:
Hm, then I will need to figure out how this works in MLIR. I first tried to go replace `!torch.vtensor<[3],!torch.qint8>`
by an mlir::Tuple `(tensor<[3],int8>, int8, f32, int8)` in `ValueTensorType::toBuiltinTensor`, but there seem to be no operations on an mlir::Tuple like insert_element or extract_element, see [1].
Then I tried replacing the type in a 1-n fashion using `typeConverter.addConversion([](Torch::ValueTensorType type, SmallVectorImpl<Type> &Tys)`, but there is no support for mapping then 1-n values in the rewriter infra [2]; the workaround being tuple unpacking.

I'm afraid that implementing this would take a while.

[1] https://mlir.llvm.org/docs/Rationale/Rationale/#tuple-types
[2] https://discourse.llvm.org/t/replacing-a-result-value-with-multiple-values-error/3221/2

Reply:
Yes, the dialect conversion framework is notoriously difficult to use. I even gave a talk about it at one point: [slides](https://drive.google.com/file/d/1FVbzCXxZzS9LBLuvpPNLWJD-XDkt54ky/view?usp=sharing) - [recording](https://drive.google.com/file/d/1VfVajitgf8ZPnd-HRkJvaJiFLhBsluXN/view?usp=sharing)

This use case is supported by the dialect conversion framework though. Doing this with a tuple type is not the right way. Registering a direct 1-n conversion is the correct way. You can find guidance in the SCF dialect's StructuralTypeConversions.cpp. In particular the manual calling of ` typeConverter->materializeSourceConversion` for 1-n conversions here: https://github.com/llvm/llvm-project/blob/fefe655baafb9aa11ae3e2a34b19aef1f47e2b8d/mlir/lib/Dialect/SCF/Transforms/StructuralTypeConversions.cpp#L85 You can ask on #mlir on the llvm discord for further guidance.

Reply:
It seems that `quantized.add` is not supported on CPU backends. I don't know how I can make end2end tests work with that.
```
class TestModule(torch.nn.Module):
  def forward(self, x, y):
    a = torch.quantize_per_tensor(x, 1.0, 0, torch.qint8)
    b = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
    c = torch.ops.quantized.add(a, y, 1.0, 0)
    return c

args = (torch.FloatTensor([0, 1]),torch.FloatTensor([0, 1]))

model = TestModule()
print(model(*args))
```
gives
```
NotImplementedError: Could not run 'aten::qscheme' with arguments from the 'CPU' backend. This could be because the operator doesn't exist for this backend, ...
```